File: Backend/app/schemas/task.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from .user import UserReference
from app.models.enums import TaskStatus, TaskPriority


class TaskUsedComponentIn(BaseModel):
    component_id: int
    quantity: float

class TaskCreate(BaseModel):
    title: str
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    completion_notes: Optional[str] = None
    assigned_to: Optional[int] = None
    asset_id: Optional[int] = None
    component_id: Optional[int] = None  # Reemplaza machine_id
    workorder_id: Optional[int] = None

class TaskRead(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    status: TaskStatus
    priority: TaskPriority
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    actual_hours: Optional[float] = None
    completion_notes: Optional[str] = None
    assigned_to: Optional[int] = None
    asset_id: Optional[int] = None
    component_id: Optional[int] = None  # Reemplaza machine_id
    workorder_id: Optional[int] = None
    created_by_id: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

class TaskReadWithUsers(BaseModel):
    id: int
    title: str
    description: Optional[str] = None
    status: TaskStatus
    priority: TaskPriority
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    actual_hours: Optional[float] = None
    completion_notes: Optional[str] = None
    assigned_to: Optional[int] = None
    assignee: Optional[UserReference] = None
    asset_id: Optional[int] = None
    component_id: Optional[int] = None  # Reemplaza machine_id
    workorder_id: Optional[int] = None
    created_by_id: int
    creator: Optional[UserReference] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    due_date: Optional[datetime] = None
    estimated_hours: Optional[float] = None
    actual_hours: Optional[float] = None
    completion_notes: Optional[str] = None
    assigned_to: Optional[int] = None
    asset_id: Optional[int] = None
    component_id: Optional[int] = None  # Reemplaza machine_id
    workorder_id: Optional[int] = None
    used_components: Optional[List[TaskUsedComponentIn]] = None


class TaskCompleteItemIn(BaseModel):
    component_id: int
    quantity: float


class TaskCompleteRequest(BaseModel):
    notes: Optional[str] = None  # alias para completion_notes
    description: Optional[str] = None
    actual_hours: Optional[float] = None
    used_components: Optional[List[TaskCompleteItemIn]] = None

File: Frontend/zyro-web/src/app/types/task.ts
/**
 * Tipos para tareas
 */

// Enums específicos para tareas
export enum TaskStatus {
  PENDING = 'PENDING',
  IN_PROGRESS = 'IN_PROGRESS',
  COMPLETED = 'COMPLETED',
  CANCELLED = 'CANCELLED'
}

export enum TaskPriority {
  LOW = 'LOW',
  MEDIUM = 'MEDIUM',
  HIGH = 'HIGH',
  CRITICAL = 'CRITICAL'
}

export interface TaskCreate {
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  due_date?: string;
  estimated_hours?: number;
  completion_notes?: string;
  assigned_to?: number;
  asset_id?: number;
  component_id?: number;
  workorder_id?: number;
}

export interface TaskRead {
  id: number;
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  due_date?: string;
  estimated_hours?: number;
  actual_hours?: number;
  completion_notes?: string;
  assigned_to?: number;
  asset_id?: number;
  component_id?: number;
  workorder_id?: number;
  created_by_id: number;
  created_at: string;
  updated_at?: string;
}

export interface TaskUsedComponentIn {
  component_id: number;
  quantity: number;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  due_date?: string;
  estimated_hours?: number;
  actual_hours?: number;
  completion_notes?: string;
  assigned_to?: number;
  asset_id?: number;
  component_id?: number;
  workorder_id?: number;
  // Al actualizar (por ejemplo al completar), se pueden enviar componentes usados
  used_components?: TaskUsedComponentIn[];
}
